### PR TITLE
fix(restore): #FOR-753 update controller and service with the use of Form Models instead of jsonObjects

### PR DESCRIPTION
**Pull request by Gregprod33**
## Describe your changes
Fixing some issues when trying to iterate through a jsonArray which contains a list of Models due to a modification of listByIds method in july. The consequence was a server exception error.
We change the way to iterate through this List of Models Form to set a property.
Also we modify the method updateMultiple in service to fit with the models and return a future <JsonArray> if succeed.
## Checklist tests
Try to delete a form and then to restore it
## Issue ticket number and link
[FOR-753](https://jira.support-ent.fr/browse/FOR-753)
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [ ] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultFormService.java [hunk 1]
```diff
@@ -481,14 +481,15 @@ public void update(String formId, JsonObject form, Handler<Either<String, JsonOb
     }
 
     @Override
-    public void updateMultiple(JsonArray forms, Handler<Either<String, JsonArray>> handler) {
+    public Future<JsonArray> updateMultiple(List<Form> forms) {
```
Comment (qperie) [inline formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultFormService.java:484]: Faire en sorte d'en faire une Future<List<Form>>

File: formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultFormService.java [hunk 2]
```diff
@@ -481,14 +481,15 @@ public void update(String formId, JsonObject form, Handler<Either<String, JsonOb
...
+        Promise<JsonArray> promise = Promise.promise();
         if (!forms.isEmpty()) {
             SqlStatementsBuilder s = new SqlStatementsBuilder();
 
             s.raw(TRANSACTION_BEGIN_QUERY);
```
Comment (qperie) [inline formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultFormService.java:489]: Utiliser les TransactionElement, le TransactionHelper etc.
Il y a un très bon exemple avec le QuestionService.update

**Review by qperie — CHANGES_REQUESTED** [review]

File: formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultFormService.java [hunk 3]
```diff
@@ -498,40 +501,45 @@ public void updateMultiple(JsonArray forms, Handler<Either<String, JsonArray>> h
...
-            s.raw(TRANSACTION_COMMIT_QUERY);
 
-            sql.transaction(s.build(), SqlResult.validResultsHandler(handler));
         }
         else {
-            handler.handle(new Either.Right<>(new JsonArray()));
+            promise.fail("[Formulaire@DefaultFormService::updateMultiple] Empty Forms List, nothing to update");
+            promise.complete(new ArrayList<>());
         }
+        String errorMessage = "[Formulaire@DefaultFormService::updateMultiple] Fail to update this forms";
```
Comment (qperie) [inline formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultFormService.java:538]:
```suggestion
        String errorMessage = "[Formulaire@DefaultFormService::updateMultiple] Fail to update these forms";
```

File: formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultFormService.java [hunk 4]
```diff
@@ -498,40 +501,45 @@ public void updateMultiple(JsonArray forms, Handler<Either<String, JsonArray>> h
...
+                transactionElements.add(new TransactionElement(query, params));
+                transactionElements.add(new TransactionElement(getUpdateDateModifFormRequest(), getParamsForUpdateDateModifFormRequest(formId.toString())));
             }
-            s.raw(TRANSACTION_COMMIT_QUERY);
 
-            sql.transaction(s.build(), SqlResult.validResultsHandler(handler));
         }
         else {
-            handler.handle(new Either.Right<>(new JsonArray()));
+            promise.fail("[Formulaire@DefaultFormService::updateMultiple] Empty Forms List, nothing to update");
```
Comment (qperie) [inline formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultFormService.java:535]:
```suggestion
            promise.fail("[Formulaire@DefaultFormService::updateMultiple] Empty forms list, nothing to update");
```

File: common/src/main/java/fr/openent/form/helpers/DateHelper.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        }
+        // No format working => is already in DAY Format
+        return inputDateStr;
+    }
+
+
+    public static Date formatDateToModel(String property, String format) {
+            SimpleDateFormat dateFormatter = new SimpleDateFormat(format);
+            Date result = new Date();
+        try {
```
Comment (qperie) [inline common/src/main/java/fr/openent/form/helpers/DateHelper.java:58]:
```suggestion
        SimpleDateFormat dateFormatter = new SimpleDateFormat(format);
        Date result = new Date();
        try {
```

File: common/src/main/java/fr/openent/form/helpers/DateHelper.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            } catch (DateTimeParseException e) {
+                // Try next format
+            }
+        }
+        // No format working => is already in DAY Format
+        return inputDateStr;
+    }
+
+
+    public static Date formatDateToModel(String property, String format) {
```
Comment (qperie) [inline common/src/main/java/fr/openent/form/helpers/DateHelper.java:55]: property c'est une date donc tu peux utiliser un nom de variable plus parlant comme les fonctions au-dessus avec inputDateStr par exemple

File: formulaire/src/main/java/fr/openent/formulaire/controllers/FormController.java [hunk 7]
```diff
@@ -782,11 +783,15 @@ public void restore(HttpServerRequest request) {
...
-                            }
-
-                            formService.updateMultiple(forms, arrayResponseHandler(request));
+                            List<Form> formList = forms.getList();
+                            formList.forEach(f -> f.setArchived(false));
+
+                            formService.updateMultiple(formList)
+                                    .onSuccess(result -> renderJson(request, new JsonArray(result)))
+                                    .onFailure(err -> {
+                                        log.error("[Formulaire@FormController::restore] Failed to restore forms" + err.getMessage());
```
Comment (qperie) [inline formulaire/src/main/java/fr/openent/formulaire/controllers/FormController.java:792]:
```suggestion
                                        log.error("[Formulaire@FormController::restore] Failed to restore forms : " + err.getMessage());
```

File: formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultFormService.java [hunk 8]
```diff
@@ -481,14 +485,13 @@ public void update(String formId, JsonObject form, Handler<Either<String, JsonOb
...
+        Promise<List<Form>> promise = Promise.promise();
+        List<TransactionElement> transactionElements = new ArrayList<>();
         if (!forms.isEmpty()) {
-            SqlStatementsBuilder s = new SqlStatementsBuilder();
-
-            s.raw(TRANSACTION_BEGIN_QUERY);
             for (int i = 0; i < forms.size(); i++) {
-                JsonObject form = forms.getJsonObject(i);
-                int formId = form.getInteger(ID, null);
+                Form form = forms.get(i);
```
Comment (qperie) [inline formulaire/src/main/java/fr/openent/formulaire/service/impl/DefaultFormService.java:493]: Tu peux faire un for each ce sera plus simple

**Review by FlorentMr — APPROVED** [review]

**Pull request merged**